packages/js/components/src/experimental-tree-control/linked-tree-utils.ts

```diff
@@ -0,0 +1,207 @@
+		);
+		linkedTree.data.isExpanded =
+			linkedTree.children.length === 0
+				? true
+				: shouldItemBeExpanded( linkedTree, createValue );
+		return linkedTree;
+	} );
+}
+
+function populateIndexes(
+	linkedTree: LinkedTree[],
+	startCount = 0
+): LinkedTree[] {
+	let count = startCount;
+
+	function populate( tree: LinkedTree[] ): number {
+		for ( const node of tree ) {
+			node.index = count;
+			count++;
+			if ( node.children ) {
+				count = populate( node.children );
+			}
+		}
+		return count;
+	}
+
+	populate( linkedTree );
+	return linkedTree;
+}
+
+export function getLinkedTree(
+	items: Item[],
+	value: string | undefined
+): LinkedTree[] {
+	const augmentedItems = items.map( ( i ) => ( {
+		...i,
+		isExpanded: false,
+	} ) );
+	return populateIndexes(
+		findChildren( augmentedItems, {}, undefined, value )
+	);
+}
+
+export function toggleNode(
+	tree: LinkedTree[],
+	number: number,
+	value: boolean
+): LinkedTree[] {
+	return tree.map( ( node ) => {
+		return {
+			...node,
+			children: node.children
+				? toggleNode( node.children, number, value )
+				: node.children,
+			data: {
+				...node.data,
+				isExpanded:
+					node.index === number ? value : node.data.isExpanded,
+			},
+			...( node.parent
+				? {
+						parent: {
+							...node.parent,
+							data: {
+								...node.parent.data,
+								isExpanded:
+									node.parent.index === number
+										? value
+										: node.parent.data.isExpanded,
+							},
+						},
+				  }
+				: {} ),
+		};
+	} );
+}
+
+// This function gets the index of the next/previous visible node in the linked tree
+export function getVisibleNodeIndex(
+	tree: LinkedTree[],
+	highlightedIndex: number,
+	direction: 'up' | 'down'
+): number | undefined {
+	if ( direction === 'down' ) {
+		for ( const node of tree ) {
+			if ( ! node.parent || node.parent.data.isExpanded ) {
+				if (
+					node.index !== undefined &&
+					node.index >= highlightedIndex
+				) {
+					return node.index;
+				}
+				const visibleNodeIndex = getVisibleNodeIndex(
+					node.children,
+					highlightedIndex,
+					direction
+				);
+				if ( visibleNodeIndex !== undefined ) {
+					return visibleNodeIndex;
+				}
+			}
+		}
+	} else {
+		for ( let i = tree.length - 1; i >= 0; i-- ) {
+			const node = tree[ i ];
+			if ( ! node.parent || node.parent.data.isExpanded ) {
+				const visibleNodeIndex = getVisibleNodeIndex(
+					node.children,
+					highlightedIndex,
+					direction
+				);
+				if ( visibleNodeIndex !== undefined ) {
+					return visibleNodeIndex;
+				}
+				if (
+					node.index !== undefined &&
+					node.index <= highlightedIndex
+				) {
+					return node.index;
+				}
+			}
+		}
+	}
+
+	return undefined;
+}
+
+export function countNumberOfItems( linkedTree: LinkedTree[] ) {
+	let count = 0;
+	for ( const node of linkedTree ) {
+		count++;
+		if ( node.children ) {
+			count += countNumberOfItems( node.children );
+		}
+	}
+	return count;
+}
+
+export function getNodeDataByIndex(
```

Comment:
Given these are exported functions it would be great if we can add some JSdocs to them.

Reply:
These won't make it to the exported package, they are just exported to be used from the other files. I've added additional comments anyway.

Reply:
Yeah I don't mean exported by the package, but exported internally.
I think its still beneficial to add JSDoc comments to these functions as it provides additional clarity even if we revisit this in a couple months. These functions are also used as utility functions so they could easily be used across the package.
It does help that this is already in TypeScript so the types will help.